self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryContracts.java

```diff
@@ -22,43 +22,392 @@
+            .anyMatch(c -> c.contributor().contributorId() == contributorId
+                && c.project().projectId() == projectId);
+        if (!contractExists) {
+            final ProjectInternal project = findOrCreateProjectInternal(
+                projectId);
+            final ContributorInternal contributor =
+                findOrCreateContributor(contributorId);
+            final ContractInternal contract =
+                new ContractInternal(project, contributor);
+            //store the contract
+            contracts.add(contract);
+            //link this contract to its project and contributor
+            project.addContract(contract);
+            contributor.addContract(contract);
+        }
+    }
+
+    /**
+     * Returns a project by id from the projects set.
+     * If the project is not found, a project will be created and added to set.
+     *
+     * @param projectId Project's
+     * @return Internal representation of a Project
+     */
+    private ProjectInternal findOrCreateProjectInternal(final int projectId) {
+        ProjectInternal projectInternal = projects.stream()
+            .filter(p -> p.projectId() == projectId)
+            .findFirst()
+            .orElse(null);
+        if (projectInternal == null) {
+            Project projectFromStorage = StreamSupport
+                    .stream(storage.projects().spliterator(), false)
+                    .filter(p -> p.projectId() == projectId)
+                    .findFirst()
+                    .orElse(null);
+            if(projectFromStorage == null){
+                throw new IllegalStateException("Project with id: "
+                    +projectId + " was not found in storage");
+            }
+            projectInternal = new ProjectInternal(projectFromStorage, this);
+            projects.add(projectInternal);
+        }
+
+        return projectInternal;
     }
+
+    /**
+     * Returns a contributor by id from the contributors set.
+     * If the contributor is not found, a contributor will
+     * be created and added to set.
+     *
+     * @param cbId Contributor's id
+     * @return Internal representation of a Contributor
+     */
+    private ContributorInternal findOrCreateContributor(final int cbId) {
+        ContributorInternal contributor = contributors.stream()
+            .filter(c -> c.contributorId() == cbId)
+            .findFirst()
+            .orElse(null);
+        if (contributor == null) {
+            contributor = new ContributorInternal(cbId, this);
+            contributors.add(contributor);
+        }
+        return contributor;
+    }
+
+
+    /**
+     * In memory implementation of a {@link Project}.
+     * Wraps a real implementation of a Project.
+     * <br>
+     * Aids {@link InMemoryContracts}
+     * <br>
+     * Has capability to link a contract.
+     * See: {@link ProjectInternal#addContract(Contract)}.
+     */
+    private static final class ProjectInternal implements Project {
```

Comment:
@criske I think these Internal classes are dummy classes to be used until we have some concrete implementations of Project and Contributor, right? :D